apollo-ios/Sources/Apollo/GraphQLResult.swift

```diff
@@ -5,34 +5,95 @@ import ApolloAPI
 /// Represents the result of a GraphQL operation.
 public struct GraphQLResult<Data: RootSelectionSet> {
 
+  /// Represents source of data
+  public enum Source: Hashable {
+    case cache
+    case server
+  }
+
   /// The typed result data, or `nil` if an error was encountered that prevented a valid response.
   public let data: Data?
   /// A list of errors, or `nil` if the operation completed without encountering any errors.
   public let errors: [GraphQLError]?
   /// A dictionary which services can use however they see fit to provide additional information to clients.
   public let extensions: [String: AnyHashable]?
-
-  /// Represents source of data
-  public enum Source: Hashable {
-    case cache
-    case server
-  }
   /// Source of data
   public let source: Source
 
   let dependentKeys: Set<CacheKey>?
 
-  public init(data: Data?,
-              extensions: [String: AnyHashable]?,
-              errors: [GraphQLError]?,
-              source: Source,
-              dependentKeys: Set<CacheKey>?) {
+  public init(
+    data: Data?,
+    extensions: [String: AnyHashable]?,
+    errors: [GraphQLError]?,
+    source: Source,
+    dependentKeys: Set<CacheKey>?
+  ) {
     self.data = data
     self.extensions = extensions
     self.errors = errors
     self.source = source
     self.dependentKeys = dependentKeys
   }
+
+  func merging(_ incrementalResult: IncrementalGraphQLResult) throws -> GraphQLResult<Data> {
+    let mergedDataDict = try merge(
+      incrementalResult.data?.__data,
+      into: self.data?.__data
+    ) { currentDataDict, incrementalDataDict in
+      try currentDataDict.merging(incrementalDataDict, at: incrementalResult.path)
+    }
+    var mergedData: Data? = nil
+    if let mergedDataDict {
+      mergedData = Data(_dataDict: mergedDataDict)
+    }
+
+    let mergedErrors = try merge(
+      incrementalResult.errors,
+      into: self.errors
+    ) { currentErrors, incrementalErrors in
+      currentErrors + incrementalErrors
+    }
+
+    let mergedExtensions = try merge(
+      incrementalResult.extensions,
+      into: self.extensions
+    ) { currentExtensions, incrementalExtensions in
+      currentExtensions.merging(incrementalExtensions) { _, new in new }
+    }
+
+    let mergedDependentKeys = try merge(
+      incrementalResult.dependentKeys,
+      into: self.dependentKeys
+    ) { currentDependentKeys, incrementalDependentKeys in
+      currentDependentKeys.union(incrementalDependentKeys)
+    }
+
+    return GraphQLResult(
+      data: mergedData,
+      extensions: mergedExtensions,
+      errors: mergedErrors,
+      source: source,
+      dependentKeys: mergedDependentKeys
+    )
+  }
+
+  fileprivate func merge<T>(
+    _ newValue: T?,
+    into currentValue: T?,
+    onCurrent: (_ currentValue: T, _ newValue: T) throws -> T
```

Comment:
What does the parameter name `onCurrent` mean? I'm not able to understand what the closure I should be passing into this parameter is supposed to be if I were calling this function.

Reply:
This is just poor follow-up on my part after refactoring. `onMerge` might be more appropriate since the closure is what gets called when there is both a new and current value.